src/Tasks/SystemState.cs

```diff
@@ -631,18 +632,18 @@ internal void SerializePrecomputedCache(string stateFile, TaskLoggingHelper log)
         /// </summary>
         /// <param name="path"></param>
         /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        /// <returns>The list of directories from the specified path.</returns>
+        private IEnumerable<string> GetDirectories(string path, string pattern)
         {
             // Only cache the *. pattern. This is by far the most common pattern
             // and generalized caching would require a call to Path.Combine which
             // is a string-copy.
             if (pattern == "*")
             {
-                instanceLocalDirectories.TryGetValue(path, out string[] cached);
+                instanceLocalDirectories.TryGetValue(path, out IEnumerable<string> cached);
                 if (cached == null)
                 {
-                    string[] directories = getDirectories(path, pattern);
+                    IEnumerable<string> directories = getDirectories(path, pattern);
                     instanceLocalDirectories[path] = directories;
```

Comment:
Do we really want to cache enumerable as is? It works now as Directory.EnumerateDirectories returns state machine that performs the search on each enumeration, but it is not guaranteed, plus it's unnecessary I/Os.

Should we rather wrap with our own type, that for array just stores as is, and for IEnumerable, performs the enumeration of target just on the first call, but stores the result for the subsequent calls?

Reply:
Other option is to expose and bubble-up a 'bool HasAnyDirectory(string path, string pattern)' method

Reply:
discussed with @rokonec offline and get back to string[]